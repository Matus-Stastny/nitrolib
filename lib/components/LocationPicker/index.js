"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _reactRelay = require("react-relay");

var _Alert = _interopRequireDefault(require("@kiwicom/orbit-components/lib/Alert"));

var _InputField = _interopRequireDefault(require("@kiwicom/orbit-components/lib/InputField"));

var _environment = _interopRequireDefault(require("../../services/environment"));

var _PickerDropDown = _interopRequireDefault(require("./primitives/PickerDropDown"));

var _NoResult = _interopRequireDefault(require("./primitives/NoResult"));

var _ClickOutside = _interopRequireDefault(require("../ClickOutside"));

var _Text = _interopRequireDefault(require("../Text"));

var _LocationPickerResultList = _interopRequireDefault(require("./components/LocationPickerResultList"));

var _placeholder = _interopRequireDefault(require("./services/placeholder"));

var _graphql;

var _ref =
/*#__PURE__*/
React.createElement(_Alert.default, {
  type: "critical"
}, React.createElement(_Text.default, {
  t: "common.api_error"
}));

var _ref2 =
/*#__PURE__*/
React.createElement(_NoResult.default, null, React.createElement(_Text.default, {
  t: "forms.places_no_results"
}));

var LocationPicker =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(LocationPicker, _React$Component);

  function LocationPicker() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, LocationPicker);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(LocationPicker)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "state", {
      active: false,
      input: ""
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "node", React.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "handleClose", function () {
      _this.setState({
        active: false
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "handleSelect", function (item) {
      var onChange = _this.props.onChange;
      onChange(item);

      _this.setState({
        active: false
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "handleChange", function (ev) {
      _this.setState({
        input: ev.target.value,
        active: true
      });
    });
    return _this;
  }

  (0, _createClass2.default)(LocationPicker, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          value = _this$props.value,
          label = _this$props.label,
          icon = _this$props.icon,
          environment = _this$props.environment;
      var _this$state = this.state,
          active = _this$state.active,
          input = _this$state.input;
      var placeholder = value ? (0, _placeholder.default)(value) : "";
      return React.createElement(_ClickOutside.default, {
        active: active,
        onClickOutside: this.handleClose
      }, React.createElement(React.Fragment, null, React.createElement(_InputField.default, {
        inlineLabel: true,
        label: label,
        placeholder: placeholder,
        onChange: this.handleChange,
        prefix: icon,
        value: active ? input : ""
      }), input && active && React.createElement(_reactRelay.QueryRenderer, {
        environment: environment,
        query: _graphql || (_graphql = function _graphql() {
          var node = require("./__generated__/LocationPickerQuery.graphql");

          if (node.hash && node.hash !== "2d45cc81c784d75b8c943feb47b77b9e") {
            console.error("The definition of 'LocationPickerQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data.");
          }

          return require("./__generated__/LocationPickerQuery.graphql");
        }),
        variables: {
          input: input
        },
        render: function render(res) {
          if (res.error) {
            return _ref;
          }

          if (!res.props) {
            return null;
          }

          if (!res.props.allLocations) {
            // TODO render this in the list if length is 0
            return _ref2;
          }

          return React.createElement(_PickerDropDown.default, {
            ref: _this2.node
          }, React.createElement(_LocationPickerResultList.default, {
            list: res.props.allLocations,
            selectedId: value && value.id,
            onSelect: _this2.handleSelect
          }));
        }
      })));
    }
  }]);
  return LocationPicker;
}(React.Component);

(0, _defineProperty2.default)(LocationPicker, "defaultProps", {
  environment: _environment.default
});
var _default = LocationPicker;
exports.default = _default;