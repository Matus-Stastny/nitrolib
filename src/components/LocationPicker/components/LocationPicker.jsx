// @flow strict
import * as React from "react";
import { graphql, QueryRenderer } from "react-relay";
import Alert from "@kiwicom/orbit-components/lib/Alert";
import InputField from "@kiwicom/orbit-components/lib/InputField";

import environmentReal from "../../../services/environment";
import PickerDropDown from "../primitives/PickerDropDown";
import ClickOutside from "../../ClickOutside";
import Text from "../../Text";
import NoResult from "../primitives/NoResult";
import LocationPickerResultList from "./LocationPickerResultList";
import type { LocationPickerRow_item } from "./__generated__/LocationPickerRow_item.graphql";
import placeholderFunc from "../services/placeholder";

type Props = {|
  label: string,
  icon: React.Node,
  error: string,
  placeholder: string,
  environment: typeof environmentReal,
|};

type State = {|
  active: boolean,
  selectedIndex: number,
  placeholderChange: string,
  results: Array<LocationPickerRow_item>,
|};

class LocationPicker extends React.Component<Props, State> {
  static defaultProps = {
    environment: environmentReal,
  };

  state = {
    selectedIndex: 0,
    active: false,
    results: [],
    // eslint-disable-next-line react/destructuring-assignment
    placeholderChange: this.props.placeholder,
  };

  node: { current: any | HTMLDivElement } = React.createRef();

  handleClose = () => {
    this.setState({
      active: false,
    });
  };

  handleResultList = (arr: Array<LocationPickerRow_item>) => {
    this.setState({
      results: arr,
    });
  };

  handleSelect = (item: LocationPickerRow_item, index: number) => {
    const { type, country, code, name } = item;
    this.setState({
      active: false,
      placeholderChange: placeholderFunc({ type, country, code, name }),
      selectedIndex: index,
    });
  };

  handleChange = (ev: SyntheticInputEvent<HTMLInputElement>) => {
    const { value } = ev.target;

    this.setState({
      placeholderChange: value,
      active: true,
    });
  };

  handleSelectItem = (ev: SyntheticKeyboardEvent<EventTarget>) => {
    ev.preventDefault();
    const { selectedIndex, results } = this.state;
    const rowHeight = this.node.current?.children[0].clientHeight;

    const ARROW_UP = 38;
    const ARROW_DOWN = 40;
    const ESC = 27;
    const ENTER = 13;

    if (ev.keyCode === ENTER) {
      this.handleSelect(results[selectedIndex], selectedIndex);
    }

    if (ev.keyCode === ESC) {
      this.handleClose();
    }

    if (ev.keyCode === ARROW_UP) {
      if (selectedIndex > 0) {
        this.setState({ selectedIndex: selectedIndex - 1 });
        this.node.current.scrollTop -= rowHeight;
      }
    }

    if (ev.keyCode === ARROW_DOWN) {
      if (selectedIndex < results.length - 1) {
        this.setState({ selectedIndex: selectedIndex + 1 });
        this.node.current.scrollTop += rowHeight;
      }
    }
  };

  render() {
    const { label, icon, error, environment, placeholder } = this.props;
    const { active, selectedIndex, placeholderChange } = this.state;

    return (
      <ClickOutside active={active} onClickOutside={this.handleClose}>
        <>
          <InputField
            label={label}
            inlineLabel
            placeholder={placeholderChange || placeholder}
            onChange={this.handleChange}
            onKeyUp={this.handleSelectItem}
            prefix={icon}
            value={active ? placeholderChange : ""}
            error={error}
          />
          {placeholderChange && active && (
            <QueryRenderer
              environment={environment}
              query={graphql`
                query LocationPickerQuery($input: String!) {
                  allLocations(last: 50, search: $input) {
                    ...LocationPickerResultList_list
                  }
                }
              `}
              variables={{ input: placeholderChange }}
              render={res => {
                if (res.error) {
                  return (
                    <Alert type="critical">
                      <Text t="common.api_error" />
                    </Alert>
                  );
                }

                if (!res.props || !res.props.allLocations) {
                  return (
                    <NoResult>
                      <Text t="forms.places_no_results" />
                    </NoResult>
                  );
                }

                return (
                  <PickerDropDown ref={this.node}>
                    <LocationPickerResultList
                      list={res.props.allLocations}
                      handleSelect={this.handleSelect}
                      selectedIndex={selectedIndex}
                      handleResults={this.handleResultList}
                    />
                  </PickerDropDown>
                );
              }}
            />
          )}
        </>
      </ClickOutside>
    );
  }
}

export default LocationPicker;
