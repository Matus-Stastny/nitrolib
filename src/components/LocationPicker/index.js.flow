// @flow strict
import * as React from "react";
import type { Environment } from "react-relay";

import type { Location } from "../../records/Location";

type Props = {|
  value: Location | null,
  onChange: (loc: Location) => void,
  label: string,
  icon: React.Node,
  // defaulted
  environment?: Environment,
|};

declare export default React.ComponentType<Props>;
