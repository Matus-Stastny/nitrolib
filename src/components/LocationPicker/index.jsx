// @flow strict
import * as React from "react";
import { graphql, QueryRenderer } from "react-relay";
import Alert from "@kiwicom/orbit-components/lib/Alert";
import InputField from "@kiwicom/orbit-components/lib/InputField";
import type { Environment } from "react-relay";

import environmentReal from "../../services/environment";
import PickerDropDown from "./primitives/PickerDropDown";
import NoResult from "./primitives/NoResult";
import ClickOutside from "../ClickOutside";
import Text from "../Text";
import LocationPickerResultList from "./components/LocationPickerResultList";
import getPlaceholder from "./services/placeholder";
import type { Location } from "../../records/Location";

// FIXME add arrow handling somehow neatly
// const ARROW_UP = 38;
// const ARROW_DOWN = 40;
// const ESC = 27;
// const ENTER = 13;

// FIXME @viktr solve these:
// centralize styles with IataPicker

// FIXME @oreqizer solve these:
// arrow handling

type Props = {|
  value: Location | null,
  onChange: (loc: Location) => void,
  label: string,
  icon?: React.Node,
  // defaulted
  environment: Environment,
|};

type State = {|
  active: boolean,
  input: string,
|};

class LocationPicker extends React.Component<Props, State> {
  static defaultProps = {
    environment: environmentReal,
  };

  state = {
    active: false,
    input: "",
  };

  node: { current: any | HTMLDivElement } = React.createRef();

  handleClose = () => {
    this.setState({
      active: false,
    });
  };

  handleSelect = (item: Location) => {
    const { onChange } = this.props;

    onChange(item);
    this.setState({
      active: false,
    });
  };

  handleChange = (ev: SyntheticInputEvent<HTMLInputElement>) => {
    this.setState({
      input: ev.target.value,
      active: true,
    });
  };

  render() {
    const { value, label, icon, environment } = this.props;
    const { active, input } = this.state;

    const placeholder = value ? getPlaceholder(value) : "";

    return (
      <ClickOutside active={active} onClickOutside={this.handleClose}>
        <>
          <InputField
            inlineLabel
            label={label}
            placeholder={placeholder}
            onChange={this.handleChange}
            prefix={icon}
            value={active ? input : ""}
          />
          {input && active && (
            <QueryRenderer
              environment={environment}
              query={graphql`
                query LocationPickerQuery($input: String!) {
                  allLocations(last: 50, search: $input) {
                    ...LocationPickerResultList_list
                  }
                }
              `}
              variables={{ input }}
              render={res => {
                if (res.error) {
                  return (
                    <Alert type="critical">
                      <Text t="common.api_error" />
                    </Alert>
                  );
                }

                if (!res.props) {
                  return null;
                }

                if (!res.props.allLocations) {
                  // TODO render this in the list if length is 0
                  return (
                    <NoResult>
                      <Text t="forms.places_no_results" />
                    </NoResult>
                  );
                }

                return (
                  <PickerDropDown ref={this.node}>
                    <LocationPickerResultList
                      list={res.props.allLocations}
                      selectedId={value && value.id}
                      onSelect={this.handleSelect}
                    />
                  </PickerDropDown>
                );
              }}
            />
          )}
        </>
      </ClickOutside>
    );
  }
}

export default LocationPicker;
