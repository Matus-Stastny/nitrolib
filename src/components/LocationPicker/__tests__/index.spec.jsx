// @flow strict
import * as React from "react";
import { shallow, mount } from "enzyme";
import GpsIcon from "@kiwicom/orbit-components/lib/icons/Gps";

import LocationPicker from "..";

describe("#LocationPicker", () => {
  const ObjectLocation = {
    code: "EN",
    country: {
      code: "US",
      id: "US",
      name: "United States",
      slug: "united-states",
    },
    id: "denver_co_us",
    location: {
      lat: 39.739236,
      lng: -104.990251,
    },
    name: "Denver",
    slug: "denver-colorado-united-states",
    subdivision: {
      code: "CO",
      id: "CO_US",
      name: "Colorado",
      slug: "colorado-united-states",
    },
    type: "city",
  };

  test("render", () => {
    const wrapper = shallow(
      <LocationPicker value={ObjectLocation} onChange={jest.fn()} label="From" />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("render active", () => {
    const wrapper = shallow(<LocationPicker value={null} onChange={jest.fn()} label="From" />);

    wrapper.setState({ active: true });

    expect(wrapper).toMatchSnapshot();
  });

  test("check input value", () => {
    const wrapper = mount(<LocationPicker value={null} onChange={jest.fn()} label="From" />);
    const input = wrapper.find("input");
    input.simulate("change", { target: { value: "New Vegas" } });
    expect(input.instance().value).toBe("New Vegas");
  });

  test("With Icon", () => {
    const wrapper = shallow(
      <LocationPicker icon={<GpsIcon />} value={null} onChange={jest.fn()} label="From" />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
