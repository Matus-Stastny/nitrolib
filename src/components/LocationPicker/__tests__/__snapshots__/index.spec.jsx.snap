// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#LocationPicker With Icon 1`] = `
<ClickOutside
  active={false}
  onClickOutside={[Function]}
>
  <InputField
    inlineLabel={true}
    label="From"
    onChange={[Function]}
    placeholder=""
    prefix={<Gps />}
    value=""
  />
</ClickOutside>
`;

exports[`#LocationPicker render 1`] = `
<ClickOutside
  active={false}
  onClickOutside={[Function]}
>
  <InputField
    inlineLabel={true}
    label="From"
    onChange={[Function]}
    placeholder="Denver (United States)"
    value=""
  />
</ClickOutside>
`;

exports[`#LocationPicker render active 1`] = `
<ClickOutside
  active={true}
  onClickOutside={[Function]}
>
  <InputField
    inlineLabel={true}
    label="From"
    onChange={[Function]}
    placeholder=""
    value=""
  />
</ClickOutside>
`;
